app/models/events/review.rb

```diff
@@ -0,0 +1,90 @@
+# == Schema Information
+#
+# Table name: reviews
+#
+#  id              :bigint           not null, primary key
+#  body            :string
+#  status          :enum             default("active")
+#  created_at      :datetime         not null
+#  updated_at      :datetime         not null
+#  github_id       :integer
+#  owner_id        :bigint
+#  pull_request_id :bigint           not null
+#
+# Indexes
+#
+#  index_reviews_on_owner_id         (owner_id)
+#  index_reviews_on_pull_request_id  (pull_request_id)
+#
+# Foreign Keys
+#
+#  fk_rails_...  (owner_id => users.id)
+#  fk_rails_...  (pull_request_id => pull_requests.id)
+#
+
+module Events
+  class Review < ApplicationRecord
+    ACTIONS = %w[submitted edited dismissed].freeze
+
+    enum status: { active: 'active', removed: 'removed' }
+
+    has_many :events, as: :handleable, dependent: :destroy
+    belongs_to :owner, class_name: 'User',
+                       foreign_key: :owner_id,
+                       inverse_of: :owned_reviews
+    belongs_to :pull_request, class_name: 'Events::PullRequest',
+                              inverse_of: :reviews
+
+    validates :status, inclusion: { in: statuses.keys }
+    validates :github_id, presence: true
+
+    attr_accessor :payload
+
+    def resolve
+      return unless handleable?
+
+      handle_action
+    end
+
+    private
+
+    def handle_action
+      send(payload['action'])
```

Comment:
Is it possible to change the use of `send` by an explicit switch or hash?

```
event_action = payload['action']

if event_action == 'submitted'
   submitted
elif event_action == 'edited'
   edited
elif event_action == 'dismissed'
   dismissed
end
```

or may be

```
event_action = payload['action']

handled_action = Hash[
   'submitted' => proc{ submitted },
   'edited' => proc{ edited },
   'dismissed' => proc{ dismissed },
]

handled_action[event_action].call
```

Reply:
Is not really perfomant doing that. I am using the `send` inside a private method, so in that case it should not be a problem. Why would you prefer doing that?

Reply:
I would think that the first alternative using `ifs` is more performant than using `send`. I think the use of `send` requires at least a message send for the `:send` message and another one for the actual message.

The second option is not performant, that is true.

The reasons would be
- to not bind the naming convention of the web hook to a method name. The method name would remain the same even if the action name changes.
- to make an explicit message send for clarity. The use of `:send` makes it harder to find references for message sends.
- to make it explicit that `handle_action` maps some actions to handler methods and to make it explicit how each action is mapped to each handler method

Reply:
The `send` is the same as using `Model.method`. Asking `if` would replace the `handleable` method, which asks if the action that is comming is included in a hash that contains the event that we are able to handle, such as `pull_request`. About the name of the method, we could use aliases if you prefer. And about the understand or explicity of the code, i beleave here we are using a sort of duck typing, we is a nice feature from rails and every dynamic programming language.

Reply:
It surprises me that the use of `:send` is as efficient as a regular message send but that is an awesome optimisation.
It also surprises me that a conditional `if` is less efficient than testing an element for inclusion in a Set. I would have thought that conditionals are highly optimised whilst testing an element inclusion in a Set requires to calculate the hash value of the element and then perform an `if` to check if the hash is present in the Set.

But I think you have a point.

However I still find the line `send(payload['action'])` a bit cryptic.

How about changing a bit the class protocol for the following:

```
    def resolve
      action = payload['action']

      return unless handleable?(action)

      handle_action(action)
    end

    private

    def handle_action(actionMethod)
      send(actionMethod)
    end

    def handleable?(actionMethod)
      ACTIONS.include?(actionMethod)
    end
```

The rationale for this change is that passing the action as a parameter gives the reader of the code full context in the method `resolve` (the reader would read the question `is the action handleable?` instead of 'is handleable?':

`handleable?(action)` vs `handleable?`

and also helps to understand at a glance that the github action name is being send as a Ruby method in `send(actionMethod)`.

As a side effect it avoids the duplication of the statement `payload['action']`.

Reply:
Whenever you are not sure about a comment or doubting, you can check it by yourself by doing a script with `Benchmark`.
And yes! I am making a refactor where a service will be created to manage the payload and the GitHub workflow. This will make everything more clear.
But now I have to get this in master to make the refactor.